staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go

```diff
@@ -214,6 +240,16 @@ func (o *ExposeServiceOptions) Complete(f cmdutil.Factory, cmd *cobra.Command) e
 	return err
 }
 
+func (o *ExposeServiceOptions) Validate() error {
+	if len(o.Name) == 0 {
+		return fmt.Errorf("name must be specified")
+	}
+	if len(o.Selector) == 0 {
+		return fmt.Errorf("selector must be specified")
+	}
+	return nil
+}
```

Comment:
Please also includes this https://github.com/kubernetes/kubernetes/blob/0ed8ee6fd7169219fbeb9bca83023a980a7a36a5/staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go#L136-L141 validation

Reply:
we are checking `port` in 'runexpose()` at line `291` so are you suggestig that it should be checked in `validate()`

Reply:
yea, I think it's better to check it at `Validate()` what do you think?

Reply:
actually we are checking other things like selector,labels in `runexpose()` so I think it should be fine

Reply:
@lauchokyip if port is not mentioned then we have to take it from exposed object , so doing this validation will invalidate if port is not given . Please check from line `289`
Also on running `go test`it is actually running `validate` function in this file and result FAILS saying `selector must be specified`, is this expected behaviour, it didn't happened before

Reply:
Hi @almas33 , can you push your code to your remote branch so I can take a look at your progress over there? I am not sure if I completely understand the question.

</br>

For your first question, the `port` should be provided by the user from here 
```go
cmd.Flags().StringVar(&o.Port, "port", o.Port, i18n.T("The port that the service should serve on. Copied from the resource being exposed, if unspecified"))
``` 
and `ports` will be provided over here ( you might need to introduce a slice to hold the `ports` to the object )
```go
		// For objects that need a port, derive it from the exposed object in case a user
		// didn't explicitly specify one via --port
		if port, found := params["port"]; found && generate.IsZero(port) {
			ports, err := o.PortsForObject(info.Object)
			if err != nil {
				return cmdutil.UsageErrorf(cmd, "couldn't find port via --port flag or introspection: %v", err)
			}
			switch len(ports) {
			case 0:
				if !isHeadlessService {
					return cmdutil.UsageErrorf(cmd, "couldn't find port via --port flag or introspection")
				}
			case 1:
				params["port"] = ports[0]
			default:
				params["ports"] = strings.Join(ports, ",")
			}
		}
```
and when the user runs the program, if port is not given, `o.Validate(f, cmd)` will throw an error.
```go
if len(o.ports) == 0 { 
	 	if len(o.port) == 0 && !isHeadlessService { 
	 		return nil, fmt.Errorf("'ports' or 'port' is a required parameter") 
	 	} 
	 } 
```

</br>

For your second question, do you know which test cases are you failing?
`selector` is actually given to all the test cases by either `input` or `flags`. From what I am seeing, if `flags` is not given then the selector will be the default selector `app:go`

Reply:
Hi @almas33 , can you push your code to your remote branch so I can take a look at your progress over there? I am not sure if I completely understand the question.

</br>

For your first question, the `port` should be provided by the user from here 
```go
cmd.Flags().StringVar(&o.Port, "port", o.Port, i18n.T("The port that the service should serve on. Copied from the resource being exposed, if unspecified"))
``` 
and `ports` will be provided over here ( you might need to introduce a slice to hold the `ports` to the object )
```go
		// For objects that need a port, derive it from the exposed object in case a user
		// didn't explicitly specify one via --port
		if port, found := params["port"]; found && generate.IsZero(port) {
			ports, err := o.PortsForObject(info.Object)
			if err != nil {
				return cmdutil.UsageErrorf(cmd, "couldn't find port via --port flag or introspection: %v", err)
			}
			switch len(ports) {
			case 0:
				if !isHeadlessService {
					return cmdutil.UsageErrorf(cmd, "couldn't find port via --port flag or introspection")
				}
			case 1:
				params["port"] = ports[0]
			default:
				params["ports"] = strings.Join(ports, ",")
			}
		}
```
and when the user runs the program, if port is not given, `o.Validate(f, cmd)` will throw an error.
```go
if len(o.ports) == 0 { 
	 	if len(o.port) == 0 && !isHeadlessService { 
	 		return nil, fmt.Errorf("'ports' or 'port' is a required parameter") 
	 	} 
	 } 
```

</br>

For your second question, do you know which test cases are you failing?
`selector` is actually given to all the test cases by either `input` or `flags`. From what I am seeing, if `flags` is not given then the selector will be the default selector `app:go`

Take a look at `expose_test.go`
```go
		{
			name: "expose-service-from-service-no-selector-defined",
			args: []string{"service", "baz"},
			ns:   "test",
			calls: map[string]string{
				"GET":  "/namespaces/test/services/baz",
				"POST": "/namespaces/test/services",
			},
			input: &corev1.Service{
				ObjectMeta: metav1.ObjectMeta{Name: "baz", Namespace: "test", ResourceVersion: "12"},
				Spec: corev1.ServiceSpec{
					Selector: map[string]string{"app": "go"},
				},
			},
			flags: map[string]string{"protocol": "UDP", "port": "14", "name": "foo", "labels": "svc=test"},
			output: &corev1.Service{
				ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "", Labels: map[string]string{"svc": "test"}},
				Spec: corev1.ServiceSpec{
					Ports: []corev1.ServicePort{
						{
							Protocol:   corev1.ProtocolUDP,
							Port:       14,
							TargetPort: intstr.FromInt(14),
						},
					},
					Selector: map[string]string{"app": "go"},
				},
			},
			expected: "service/foo exposed",
			status:   200,
		},
```

Reply:
@lauchokyip  `name: "expose-service-from-service-no-selector-defined",` this test case is failing,this is the first one so I expect others are also failing
For code you can look at the latest commit , I guess I have to change tests since `expose.go` have been changed

Reply:
Does that answer your question?

Reply:
> You might need to change the `expected` result from `"service/foo exposed"` to `"error: selector must be specified"`

This doesn't seems to be working , is my `expose.go` code correct?

Reply:
Your latest commit doesn't seem to be incorparted the changes I purposed over here (https://github.com/kubernetes/kubernetes/pull/99891#issuecomment-900376032). Have you pushed your changes to the remote branch?